File: problems/003/003.hs
import Control.Exception.Base (evaluate)
import System.Clock
import System.Environment (getArgs)
import System.IO

input :: Int
input = 600851475143

solution :: Int -> Int
solution = factor 3
  where factor f 1 = f
        factor f n | f * f > n = n
                   | n `rem` f == 0 = factor f (n `div` f)
                   | otherwise = factor (f+2) n

run :: (a -> b) -> a -> Int -> IO ()
run fun inp iters
  | iters <= 0 = return ()
  | otherwise = evaluate (fun inp) >> run fun inp (iters-1)

benchmark :: (a -> b) -> a -> Int -> IO Int
benchmark fun inp iters = do
  start <- getTime Monotonic
  run fun inp iters
  end <- getTime Monotonic
  let secs = sec end - sec start
  let nsecs = nsec end - nsec start
  let elapsed = 1000000000 * secs + nsecs
  return elapsed

benchmarkLoop :: IO ()
benchmarkLoop = do
  line <- getLine
  let iters = read line
  time <- benchmark solution input iters
  print time
  hFlush stdout
  benchmarkLoop

main = do
  args <- getArgs
  case args of
    [ "-a" ] -> print $ solution input
    [] -> benchmarkLoop

File: problems/003/003.cpp
#include<cstdint>
#include<cstring>
#include<ctime>
#include<iostream>
#include<string>

using namespace std;

uint64_t solution() {
  uint64_t n = 600851475143;

  for (uint64_t factor = 3; ; factor += 2) {
    while (n % factor == 0)
      n /= factor;

    if (factor * factor > n)
      return n;
    else if (n == 1)
      return factor;
  }
}

int main(int argc, char* argv[]) {
  if (argc == 2 && strcmp(argv[1], "-a") == 0) {
    cout << solution() << endl;
    return 0;
  }

  for (string line; getline(cin, line);) {
    int iters = atoi(line.c_str());

    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC_RAW, &start);
    for (int i = 0; i < iters; i++) {
      solution();
    }
    clock_gettime(CLOCK_MONOTONIC_RAW, &end);

    long int secs = end.tv_sec - start.tv_sec;
    long int nsecs = end.tv_nsec - start.tv_nsec;
    cout << secs * 1000000000 + nsecs << endl;
  }

  return 0;
}
